akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -313,6 +317,118 @@ private[stream] class VirtualPublisher[T]() extends Publisher[T] {
 /**
  * INTERNAL API
  */
+private[stream] case class MaterializedValueSource[M](
+  shape: SourceShape[M] = SourceShape[M](new Outlet[M]("Materialized.out")),
+  attributes: OperationAttributes = OperationAttributes.name("Materialized")) extends StreamLayout.Module {
+
+  override def subModules: Set[Module] = Set.empty
+  override def withAttributes(attr: OperationAttributes): Module = this.copy(shape = amendShape(attr), attributes = attr)
+  override def carbonCopy: Module = this.copy(shape = SourceShape(new Outlet[M]("Materialized.out")))
+
+  override def replaceShape(s: Shape): Module =
+    if (s == shape) this
+    else throw new UnsupportedOperationException("cannot replace the shape of MaterializedValueSource")
+
+  def amendShape(attr: OperationAttributes): SourceShape[M] = {
+    attr.nameOption match {
+      case None ⇒ shape
+      case s: Some[String] if s == attributes.nameOption ⇒ shape
+      case Some(name) ⇒ shape.copy(outlet = new Outlet(name + ".out"))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object MaterializedValuePublisher {
+  final val NotRequested = 0
+  final val Requested = 1
+  final val Completed = 2
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] class MaterializedValuePublisher extends Publisher[Any] {
+  import MaterializedValuePublisher._
+
+  private val value = new AtomicReference[AnyRef](null)
+  private val registeredSubscriber = new AtomicReference[Subscriber[_ >: Any]](null)
+  private val requestState = new AtomicInteger(NotRequested)
+
+  private def close(): Unit = {
+    requestState.set(Completed)
+    value.set(null)
+    registeredSubscriber.set(null)
+  }
+
+  private def tryOrClose(block: ⇒ Unit): Unit = {
+    try block catch {
+      case v: ReactiveStreamsCompliance.SpecViolation ⇒
+        close()
+      // What else can we do here?
+      case NonFatal(e) ⇒
+        val sub = registeredSubscriber.get()
+        if ((sub ne null) &&
+          requestState.compareAndSet(NotRequested, Completed) || requestState.compareAndSet(Requested, Completed)) {
+          sub.onError(e)
+        }
+        close()
+        throw e
+    }
+  }
+
+  def setValue(m: Any): Unit =
+    tryOrClose {
+      if (value.compareAndSet(null, m.asInstanceOf[AnyRef]) && requestState.get() == Requested)
```

Comment:
> I.e. when both run concurrently is it guaranteed that at least one of the pushAndClose terms fire?

I think so. First they do a CAS on their "own" side and a get on the other side. The possible overlaps are (removing symmetric cases where you can relabel A->B, B->A):

- A-CAS
- A-GET
- B-CAS
- B-GET - fires here

- A-CAS
- B-CAS
- A-GET - fires here
- B-GET - fires here (pushAndClose CAS fails)

- A-CAS
- B-CAS
- B-GET - fires here
- A-GET - fires here (pushAndClose CAS fails)

The proof that there are no cases:
 - all permutations of 4 operations are 4! = 24
 - the operations of A and B are cannot be reordered, so there are 24 / (2 * 2) = 6 actual orderings
 - if we don't count cases which are a simple relabeling A->B, B->A, we get 6 / 2 = 3 reorderings which are all enumerated above. 

pushAndClose protects against double onNext by doing a CAS itself. Actually the CAS is probably not necessary on the setValue side, but I like this symmetry and it wouldn't buy much to change it to a SET.

I will add a comment about this.